src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,216 @@
+
+		auto thumbnail = Thumbnail(thumbnailSize);
+
+		for (auto
+			smallIndex = size_t{0};
+			smallIndex < thumbnail.size();
+			smallIndex += 1
+		) {
+			auto bigIndex 	= smallIndex * thumbnailSizeDivisor;
+			auto bigItStart = biggerThumbnail.cbegin() + bigIndex;
+			auto bigItEnd 	= std::min(bigItStart + thumbnailSizeDivisor, biggerThumbnail.end());
+
+			thumbnail[smallIndex] = std::accumulate(bigItStart, bigItEnd, Bit{});
+		}
+
+		m_thumbnailCache->push_back(thumbnail);
+		thumbnailSize /= thumbnailSizeDivisor;
+	}
+}
+
+void SampleThumbnail::draw(QPainter& painter, const Bit& bit, float lineX, int centerY,
+	float scalingFactor, const QColor& color, const QColor& innerColor)
+{
+	const auto lengthY1 = bit.max * scalingFactor;
+	const auto lengthY2 = bit.min * scalingFactor;
+
+	const auto lineY1 = centerY - lengthY1;
+	const auto lineY2 = centerY - lengthY2;
+
+	const auto innerLineY1 = centerY - bit.max*0.6 * scalingFactor;
+	const auto innerLineY2 = centerY - bit.min*0.6 * scalingFactor;
+
+	painter.drawLine(QPointF{lineX, lineY1}, QPointF{lineX, lineY2});
+	painter.setPen(innerColor);
+
+	painter.drawLine(QPointF{lineX, innerLineY1}, QPointF{lineX, innerLineY2});
+	painter.setPen(color);
+}
+
+void SampleThumbnail::visualize(const VisualizeParameters& parameters, QPainter& painter) const
+{
+	const auto& clipRect = parameters.clipRect;
+	const auto& sampRect = parameters.sampRect.isNull() ? clipRect : parameters.sampRect;
+	const auto& drawRect = parameters.drawRect.isNull() ? clipRect : parameters.drawRect;
+
+	const auto sampleViewLength = parameters.sampleEnd - parameters.sampleStart;
+
+	const auto x = sampRect.x();
+	const auto height = clipRect.height();
+	const auto halfHeight = height / 2;
+	const auto width = sampRect.width();
+	const auto centerY = clipRect.y() + halfHeight;
+
+	if (width < 1) { return; }
+
+	const auto scalingFactor = halfHeight * parameters.amplification;
+
+	const auto color = painter.pen().color();
+	const auto innerColor = color.lighter(123);
+
+	const auto widthSelect = static_cast<std::size_t>(static_cast<float>(width) / sampleViewLength);
+
+	auto thumbnailIt = m_thumbnailCache->end();
+
+	do
+	{
+		thumbnailIt--;
+	}
+	while (thumbnailIt != m_thumbnailCache->begin() && thumbnailIt->size() < widthSelect);
+
+	const auto& thumbnail = *thumbnailIt;
+
+	const auto thumbnailLastSample 	= std::max<size_t>(static_cast<size_t>(parameters.sampleEnd * thumbnail.size()), 1) - 1;
+	const auto tViewStart 			= static_cast<long>(parameters.sampleStart * thumbnail.size());
+	const auto tViewLast 			= std::min<size_t>(thumbnailLastSample, thumbnail.size() - 1);
+	const auto tLast				= thumbnail.size()-1;
+	const auto thumbnailViewSize 	= thumbnailLastSample + 1 - tViewStart;
+
+	const auto pixelIndexStart 	= std::max(std::max(clipRect.x(), drawRect.x()), x);
+	const auto pixelIndexEnd 	= std::min(std::min(clipRect.width(), drawRect.width()), width) + pixelIndexStart;
```

Comment:
Tip: [cppreference](https://en.cppreference.com/w/cpp/algorithm/min) is a good place to go to when you wonder if C++ already has some functionality that would make your code simpler. Always think about how to make things simpler, no matter how small.

```suggestion
	const auto pixelIndexStart 	= std::max({clipRect.x(), drawRect.x(), x});
	const auto pixelIndexEnd 	= std::min({clipRect.width(), drawRect.width(), width}) + pixelIndexStart;
```